chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,34 @@
+//! Logic for resharding flat storage in parallel to chain processing.
+//!
+//! See [FlatStorageReshard] for more details about how the resharding takes place.
+
+use std::sync::Arc;
+
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::types::BlockHeight;
+
+use crate::types::RuntimeAdapter;
+
+/// `FlatStorageReshard` takes care of updating flat storage when a resharding event
+/// happens.
+///
+/// On an high level, the operations supported are:
+/// - #### Shard splitting
+///     Parent shard must be split into two children. The entire operation freezes the flat storage
+///     for the involved shards.
+///     Children shards are created empty and the key-values of the parent will be copied into one of them,
+///     in the background.
+///
+///     After the copy is finished the children shard will have the correct state at some past block height.
+///     It'll be necessary to perform catchup before the flat storage can be put again in Ready state.
+///     The parent shard storage is not needed anymore and can be removed.
+pub struct FlatStorageReshard {
```

Comment:
Oh I meant it to be `FlatStorageResharder` taking inspiration from `FlatStorageCreator` . Still, it might very well end up being a message for some actor